plugins/notifications-backend/src/service/handlers.ts

```diff
@@ -0,0 +1,425 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/* eslint-disable func-names */
+import { CatalogClient } from '@backstage/catalog-client';
+
+import { Knex } from 'knex';
+
+import { Components, Paths } from '../openapi';
+import {
+  DefaultMessageScope,
+  DefaultOrderBy,
+  DefaultOrderDirection,
+  DefaultPageNumber,
+  DefaultPageSize,
+} from './constants';
+import { ActionsInsert, dbValToBoolean, MessagesInsert } from './db';
+import {
+  MessageScopes,
+  NotificationsFilterRequest,
+  NotificationsOrderByDirections,
+  NotificationsOrderByFields,
+  NotificationsSortingRequest,
+} from './types';
+
+// createNotification
+// returns string id of created notification
+export async function createNotification(
+  dbClient: Knex<any, any>,
+  catalogClient: CatalogClient,
+  req: Paths.CreateNotification.RequestBody,
+): Promise<Paths.CreateNotification.Responses.$200> {
+  let isUser = false;
+
+  // validate users
+  if (Array.isArray(req.targetGroups) && req.targetGroups.length > 0) {
+    isUser = true;
+    const promises = req.targetGroups.map(group => {
+      return catalogClient.getEntityByRef(`group:${group}`).then(groupRef => {
+        if (!groupRef) {
+          throw new Error(`group '${group}' does not exist`);
+        }
+      });
+    });
+
+    await Promise.all(promises);
+  }
+
+  // validate groups
+  if (Array.isArray(req.targetUsers) && req.targetUsers.length > 0) {
+    isUser = true;
+    const promises = req.targetUsers.map(user => {
+      return catalogClient.getEntityByRef(`user:${user}`).then(userRef => {
+        if (!userRef) {
+          throw new Error(`user '${user}' does not exist`);
+        }
+      });
+    });
+
+    await Promise.all(promises);
+  }
+
+  // validate actions
+  if (Array.isArray(req.actions)) {
+    req.actions.forEach(action => {
+      if (!action.title || !action.url) {
+        throw new Error('Both action title and url are mandatory.');
+      }
+    });
+  }
+
+  const row: MessagesInsert = {
+    origin: req.origin,
+    title: req.title,
+    message: req.message,
+    topic: req.topic,
+    is_system: !isUser,
+  };
+
+  let messageId: string;
+
+  const ret = dbClient.transaction(trx => {
+    return trx
+      .insert(row)
+      .returning<string, { id: string }[]>('id')
+      .into('messages')
+      .then(ids => {
+        messageId = ids[0].id;
+        if (Array.isArray(req.targetUsers)) {
+          const userInserts = req.targetUsers.map(user => {
+            return {
+              message_id: messageId,
+              user: user,
+            };
+          });
+          return trx('users').insert(userInserts);
+        }
+
+        return undefined;
+      })
+      .then(() => {
+        if (Array.isArray(req.targetGroups)) {
+          const groupInserts = req.targetGroups.map(group => {
```

Comment:
I wonder why we have group notifications? In my version, all notifications were simply targeted to user(s) so if you send a notification to a group, it will be sent to all members of that group. It also works for other entities so that if you send a notification to, let's say System, it will send the notification to the owner of that system.

Reply:
This actually would mean that when the user is removed from some group, the notifications would disappear from the user and when the user is added to some group, they would appear. Is this really wanted feature?

Reply:
groups are needed. the only real consumer we have at the moment requires it. besides, why not? backstage has entities of type group and of type user. it makes sense to have both.

Reply:
If you insist you need groups (including the functionality that users leaving groups or joining them later will see / unsee some notifications sent to that group), then sure. Even if groups are supported, is there some reason why these could not live in the same table and just use the entity ref (either user or group) as the target?